pkg/storageops/azure/azure.go

```diff
@@ -0,0 +1,808 @@
+	disksClient.AddToUserAgent(userAgentExtension)
+
+	vmsClient := compute.NewVirtualMachinesClient(subscriptionID)
+	vmsClient.Authorizer = authorizer
+	vmsClient.PollingDelay = clientPollingDelay
+	vmsClient.RetryAttempts = clientRetryAttempts
+	vmsClient.AddToUserAgent(userAgentExtension)
+
+	snapshotsClient := compute.NewSnapshotsClient(subscriptionID)
+	snapshotsClient.Authorizer = authorizer
+	snapshotsClient.PollingDelay = clientPollingDelay
+	snapshotsClient.RetryAttempts = clientRetryAttempts
+	snapshotsClient.AddToUserAgent(userAgentExtension)
+
+	return &azureOps{
+		instance:          instance,
+		resourceGroupName: resourceGroupName,
+		disksClient:       &disksClient,
+		vmsClient:         &vmsClient,
+		snapshotsClient:   &snapshotsClient,
+	}, nil
+}
+
+func (a *azureOps) Create(
+	template interface{},
+	labels map[string]string,
+) (interface{}, error) {
+	d, ok := template.(*compute.Disk)
+	if !ok {
+		return nil, storageops.NewStorageError(
+			storageops.ErrVolInval,
+			"Invalid volume template given",
+			a.instance,
+		)
+	}
+
+	ctx := context.Background()
+	future, err := a.disksClient.CreateOrUpdate(
+		ctx,
+		a.resourceGroupName,
+		*d.Name,
+		compute.Disk{
+			Location: d.Location,
+			Type:     d.Type,
+			Zones:    d.Zones,
+			Tags:     formatTags(labels),
+			Sku:      d.Sku,
+			DiskProperties: &compute.DiskProperties{
+				CreationData: &compute.CreationData{
+					CreateOption: compute.Empty,
+				},
+				DiskSizeGB: d.DiskProperties.DiskSizeGB,
+			},
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create disk: %v", err)
+	}
+
+	t := func() (interface{}, bool, error) {
+		_, err := future.Result(*a.disksClient)
+		return "", true, err
+	}
+	if _, err := task.DoRetryWithTimeout(t, resultTimeout, resultRetryInterval); err != nil {
+		return nil, err
+	}
+
+	// XXX We don't get back the expected disk object from future.Result(). Some of the
+	// fields are missing and the name is different. Hence we need to query to object again.
+	dd, err := a.disksClient.Get(context.Background(), a.resourceGroupName, *d.Name)
+	return &dd, err
+}
+
+func (a *azureOps) GetDeviceID(disk interface{}) (string, error) {
+	if d, ok := disk.(*compute.Disk); ok {
+		return *d.Name, nil
+	} else if s, ok := disk.(*compute.Snapshot); ok {
+		return *s.Name, nil
+	}
+	return "", storageops.NewStorageError(
+		storageops.ErrVolInval,
+		"Invalid volume given",
+		a.instance,
+	)
+}
+
+func (a *azureOps) Attach(diskName string) (string, error) {
+	vm, err := a.describeInstance()
+	if err != nil {
+		return "", fmt.Errorf("cannot get vm %v: %v", a.instance, err)
+	}
+
+	disk, err := a.disksClient.Get(
+		context.Background(),
+		a.resourceGroupName,
+		diskName,
+	)
+	if err != nil {
+		return "", fmt.Errorf("cannot get disk %v: %v", diskName, err)
+	}
+
+	nextLun := nextAvailableLun(*vm.StorageProfile.DataDisks)
+	if nextLun < 0 {
+		return "", fmt.Errorf("No LUN available to attach the disk")
```

Comment:
Can we return the number of attached luns here to help debug in case of issues?

Reply:
Just printing the number of data disks, which indirectly gives the same info.